avocado/plugins/assets.py

```diff
@@ -301,33 +301,92 @@ def configure(self, parser):
+                                 key='name',
+                                 help_msg=help_msg,
+                                 default=None,
+                                 key_type=str,
+                                 parser=register_subcommand_parser,
+                                 positional_arg=True)
+
+        help_msg = "Path to asset that you would like to register manually."
+        settings.register_option(section='assets.register',
+                                 key='url',
+                                 help_msg=help_msg,
+                                 default=None,
+                                 key_type=str,
+                                 parser=register_subcommand_parser,
+                                 positional_arg=True)
 
-            # check if we should ignore the errors
-            if config.get('assets.fetch.ignore_errors'):
-                exitcode = exit_codes.AVOCADO_ALL_OK
+        help_msg = "SHA1 hash of this asset."
+        settings.register_option(section='assets.register',
+                                 key='sha1_hash',
+                                 help_msg=help_msg,
+                                 default=None,
+                                 key_type=str,
+                                 metavar="SHA1",
+                                 long_arg='--hash',
+                                 parser=register_subcommand_parser)
 
+    def handle_fetch(self, config):
+        exitcode = exit_codes.AVOCADO_ALL_OK
+        # fetch assets from instrumented tests
+        for test_file in config.get('assets.fetch.references'):
+            if os.path.isfile(test_file) and test_file.endswith('.py'):
+                LOG_UI.debug('Fetching assets from %s.', test_file)
+                success, fail = fetch_assets(test_file)
+
+                for asset_file in success:
+                    LOG_UI.debug('  File %s fetched or already on'
+                                 ' cache.', asset_file)
+                for asset_file in fail:
+                    LOG_UI.error(asset_file)
+
+                if fail:
+                    exitcode |= exit_codes.AVOCADO_FAIL
+            else:
+                LOG_UI.warning('No such file or file not supported: %s',
+                               test_file)
+                exitcode |= exit_codes.AVOCADO_FAIL
+
+        # check if we should ignore the errors
+        if config.get('assets.fetch.ignore_errors'):
+            return exit_codes.AVOCADO_ALL_OK
         return exitcode
+
+    def handle_register(self, config):
+        cache_dirs = data_dir.get_cache_dirs()
+        name = config.get('assets.register.name')
+        asset_hash = config.get('assets.register.sha1_hash')
+        location = config.get('assets.register.url')
+        # Adding a twice the location is a small hack due the current logic to
+        # return "by_name". This needs to be improved soon.
+        asset = Asset(name=name,
```

Comment:
IIUC, the use case for this feature includes the ability of referencing *local* files, that is, when a file is publicly accessible, you can do:

```
$ avocado assets register fedora-33-x86_64-kernel http://distrib-coffee.ipsl.jussieu.fr/pub/linux/fedora/linux/releases/33/Everything/x86_64/os/images/pxeboot/vmlinuz 
Done. Now you can reference it by name fedora-33-x86_64-kernel
```

But suppose you have to download a firmware or have the file locally, you should be able to do:

```
$ avocado assets register fedora-31-x86_64-kernel /boot/vmlinuz-5.8.18-100.fc31.x86_64 
Unsupported protocol:
```

The following works, but I think the current approach would confuse a LOT of people:

```
$ avocado assets register fedora-31-x86_64-kernel file:///boot/vmlinuz-5.8.18-100.fc31.x86_64 
Done. Now you can reference it by name fedora-31-x86_64-kernel
```

So I advocate in favor of supporting the "non protocol" version too.

Reply:
Are you suggesting that the default protocol could be `file://`?

Reply:
That's one way to do it.  The end result is that users giving a local file path, end up getting that.  I'm not sure if there are drawbacks to defaulting to `file://`... I guess it depends on where that default will be set.

Reply:
I think it will improve usability, so I'm ok with that approach.